node/core/dispute-coordinator/src/cache.rs

```diff
@@ -0,0 +1,537 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! An LRU cache implementation for [`OverlayedBackend`].
+
+use polkadot_primitives::v2::{CandidateHash, SessionIndex};
+
+use lru_cache::LruCache;
+use std::{collections::HashMap, time::Duration};
+
+use super::db::v1::{CandidateVotes, RecentDisputes};
+use crate::backend::BackendWriteOp;
+
+/// An arbitrary LRU cache size for candidate votes.
+const CACHE_SIZE: usize = 1024;
+
+/// The write back interval in seconds
+#[cfg(not(test))]
+pub const WRITE_BACK_INTERVAL: Duration = Duration::from_secs(30);
+#[cfg(test)]
+pub const WRITE_BACK_INTERVAL: Duration = Duration::from_secs(0);
+
+/// A read-only cache (LRU) which records changes and outputs them as `BackendWriteOp` which can
+/// be fed to a `Backend` for persisting to the DB. It allows the subsystem to avoid
+/// calling into the database to retrieve or write data.
+///
+/// How it works:
+/// - Candidate votes, session and disputes are kept in memory in the LRU cache and their
+/// state is tracked (Cached, Dirty, Persisted).
+/// - Once the cache is full (`CACHE_SIZE`), we start moving evicting entries into
+/// `evicted_candidate_votes`.
+/// - If entries are modified via `update_*`, then these will be marked `Dirty`
+///
+/// How do I get the `BackendWriteOps`?
+/// - `into_write_ops()` scans the LRU cache and the evicted hashmap to construct
+/// a list of `BackendWriteOps`.
+/// - This method is supposed to be called when we want to write in-memory changes
+/// to the DB. It will consume the cache in the process, but it will return a new one
+/// with the same LRU cached values.
+
+pub struct OverlayCache {
+	cached_earliest_session: Option<PersistedCacheEntry<SessionIndex>>,
+	cached_recent_disputes: Option<PersistedCacheEntry<RecentDisputes>>,
+	cached_candidate_votes:
+		LruCache<(SessionIndex, CandidateHash), PersistedCacheEntry<Option<CandidateVotes>>>,
+
+	// All entries here need to be `Dirty`.
+	evicted_candidate_votes:
+		HashMap<(SessionIndex, CandidateHash), PersistedCacheEntry<Option<CandidateVotes>>>,
+	dirty: bool,
+	capacity: usize,
+}
+
+/// Cache entry state definitions.
+#[derive(Clone, Debug)]
+pub enum PersistedCacheEntryState {
+	/// The entry is in memory and is unmodified. This is the initial state.
+	Cached,
+	/// The entry is in memory and was modified (needs to be persisted later).
+	Dirty,
+	/// The entry is cached and has been persisted in the DB.
+	Persisted,
```

Comment:
how `Persisted` is different from `Cached`?
why do we use a state instead of keeping dirty stuff separately (ala data-oriented design)?

Reply:
Persisted is the same as `Cached`, plus the fact that we wrote it to disk at least once. I was hoping to expose the state transitions as metrics in the future if this cache is useful. So that's why I am using the extra `Persisted` state.

We do keep stuff separately, but that's just what we evict from the LRU cache. This leads to more code, as we need to lookup multiple hashmaps. I am trying to avoid this for everything else.

Reply:
I meant the following layout instead:
```rust
State {
    earliest_session: Option<SessionIndex>,
    recent_disputes: Option<RecentDisputes>,
    candidate_votes: LruCache<(SessionIndex, CandidateHash), CandidateVotes>,
}
OverlayCache {
    cached: State,
    dirty: State,
}
```

Reply:
This would work as well, but the same applies regarding complexity. The wrapper type also enables us to observe lifetime of cache entries in the future.